### PR TITLE
Fix rule of fill violations in utils and tests

### PR DESCRIPTION
**Pull request by seiyab**
## Description
https://github.com/prettier/prettier/issues/16582

Do we need changelog for change in `stripTrailingHardline()`?

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] ~(If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).~
- [x] ~(If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.~
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: tests/integration/__tests__/debug-api.js [hunk 1]
```diff
@@ -72,18 +72,6 @@ describe("API", () => {
...
-    But if we pass array to it, it will flat array
-    */
-    const cleanDoc = (parts) => parts.flat();
-    expect(
-      await formatDoc(
-        // The argument of fill must not be passed to cleanDoc because it's not a doc
-        fill(cleanDoc(["foo", literalline, "bar"])), // invalid fill
-      ),
-    ).toBe('fill(["foo", literallineWithoutBreakParent, breakParent, "bar"])');
-
```
Comment (seiyab) [inline tests/integration/__tests__/debug-api.js:86]: I'm considering this test case doesn't make sense. I'll add some tests if we need alternative tests.

**Comment by seiyab** [comment]
Oops a test case I recently added might be somewhat fragile. I'll fix it (increase `relativeMargin`) if the problem repeats.

```
FAIL tests/unit/doc-printer.js
`printDocToString` has linear time complexity at most to print fill() > numWords=20000
    Error: expect(received).toBeLessThan(expected)
    
    Expected: < 41.39885533368215
    Received:   41.98062500008382
```

**Comment by fisker** [comment]
Though we expect "fill" to have odd length, but we didn't throw for violations, we even did wrong things in core, so plugins should already made mistakes, we can't force them to follow the strict rule, at least not a good time.

I know I said add assert for "fill", but it should only run in dev mode.

For "stripTrailingHardline", instead of remove "hardline" from parts directly, maybe we can replace it with empty string. Do you think it will work?

**Comment by seiyab** [comment]
> so plugins should already made mistakes

Yes, I agree with you, they shuold.

However, maybe the comment misses an important point. We don't (won't) expect `fill()` to have odd elements. The rule we are working on is that even-th (1-indexed) elements have to be line-like doc.
What I'm worrying about `stripTrailingHardline()` is, change in behavior.
It's not easy to demonstrate the change simply. New `stripTrailingHardline()` always output valid `fill()` when input is valid `fill()`. The one in `main` can output invalid `fill()` for valid `fill()`. The behaviour will change, anyway.

**Comment by seiyab** [comment]
And I found that new logic might be able to get simpler 😅 . I was somewhat confused and it might confused you. I'll see it later.

**Comment by seiyab** [comment]
I reverted some commits to get this PR to be conservative.
We can discuss whether we let current `stripTrailingHardline()` go with https://github.com/prettier/prettier/pull/16656/commits/8565e91b8cbf67248204579632d7832fba5b6fe0 or fix `stripTrailingHardline()`, later.

**Review by fisker — APPROVED** [review]

**Comment by fisker** [comment]
seiyab@8565e91 seemsgood, but it should not need call `cleanDoc` since the parts already "clean"ed.

**Pull request merged**